src/main/java/sirius/biz/storage/layer3/VirtualFile.java

```diff
@@ -1267,6 +1270,124 @@ public boolean loadFromUrl(URL url, boolean force) throws IOException {
+    public VirtualFile resolveOrLoadChildFromURL(URL url, boolean force, Predicate<String> fileExtensionVerifier)
+            throws IOException {
+        String path = loadPathFromUrl(url, fileExtensionVerifier);
+        if (Strings.isEmpty(path)) {
+            String contentDisposition = loadContentDisposition(url);
+            if (fileExtensionVerifier.test(contentDisposition)) {
+                path = contentDisposition;
+            } else {
+                throw Exceptions.createHandled()
+                                .withNLSKey("VirtualFile.loadFromUrl.noValidPath")
+                                .set("url", url)
+                                .handle();
+            }
+        }
+        VirtualFile file = resolve(path);
+        file.loadFromUrl(url, force);
+        return file;
+    }
+
+    private String loadContentDisposition(URL url) throws IOException {
+        HttpURLConnection urlConnection = (HttpURLConnection) url.openConnection();
+        urlConnection.connect();
+        String contentDisposition = urlConnection.getHeaderField(HttpHeaderNames.CONTENT_DISPOSITION.toString());
+        urlConnection.disconnect();
+        return contentDisposition;
+    }
+
+    @Nullable
+    private String loadPathFromUrl(URL url, Predicate<String> fileExtensionVerifier) {
+        QueryStringDecoder qsd = new QueryStringDecoder(url.toString(), StandardCharsets.UTF_8);
+        return qsd.parameters()
+                  .values()
+                  .stream()
+                  .flatMap(List::stream)
+                  .filter(fileExtensionVerifier)
+                  .findFirst()
+                  .orElseGet(() -> {
+                      if (fileExtensionVerifier.test(url.getPath())) {
+                          return url.getPath();
+                      }
+                      return null;
+                  });
+    }
+
+    /**
+     * Attempts to resolve the file from the given URL or performs a download just as {@link #resolveOrLoadChildFromURL(URL, boolean, Predicate)} but reports to the given process.
+     * <p>
+     * This will increment one of the timings (downloaded or download skipped) and also directly report IO
+     * errors to the process without spamming the system logs.
+     *
+     * @param url                   the URL which determines the filename/path as well as the source of the file to fetch
+     * @param force                 if set to <tt>true</tt> a download will be performed, even if we already downloaded the file
+     *                              before and no modification was detected
+     * @param fileExtensionVerifier specifies which extensions are accepted. This should be used to prevent using
+     *                              ".php" or the like as effective file name
+     * @param processContext        the process to report to
+     * @return the file which has been resolved (and downloaded if necessarry)
+     */
+    public VirtualFile resolveOrLoadChildFromURL(URL url,
```

Comment:
I'd handle this outside of the VFS API - nearly everything can be done via 
try {
resolveOrLoadChildFromURL
} catch {
  processContext.handle
}

Reply:
we could replace the IOException by another HandledException in this case.... and the addTiming should be possible using TaskContext.addTming